messaging/chat.py

```diff
@@ -0,0 +1,66 @@
+from django.contrib.auth import get_user_model
+from django.utils import timezone
+
+from .models import Message, Thread, UserThread
+from messaging.rapidpro_client import RapidProClient
+
+User = get_user_model()
+
+
+class ChatManager:
+    def __init__(self, thread):
+        if thread is None:
+            raise Exception('No thread found.')
+        self.thread = thread
+
+    def _record_message_in_database(self, sender, rapidpro_message_id, text, quick_replies):
+        # Messages sent from User to RapidPro server don't have rapidpro_message_id
+        if rapidpro_message_id:
+            message, created = Message.objects.get_or_create(rapidpro_message_id=rapidpro_message_id, defaults={
+                'sender': sender,
+                'text': text,
+                'quick_replies': quick_replies,
+                'thread': self.thread,
+            })
+            if not created:
+                # If the message already exists, we concatenate the newly received
+                # messages with the existing message. An assumption here is that
+                # messages will always be received in the correct order. This should
+                # be confirmed with RapidPro
```

Comment:
I do not think this is guaranteed, but there's nothing we can do about it. With the message length limit set to 6400 in recent released of RapidPro, messages being split should be exceedingly rare.